### PR TITLE
Flag to use CPU on TEDPolicy train

### PR DESCRIPTION
**Pull request by WashingtonBispo**
**Proposed changes**:
- [https://github.com/RasaHQ/rasa/issues/10897](https://github.com/RasaHQ/rasa/issues/10897)

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/ted_policy.py
Comment (melindaloubser1) [inline rasa/core/policies/ted_policy.py:638]: Can this be refactored so there isn't duplication of the training calls? E.g. using a context manager

File: rasa/core/policies/ted_policy.py
Comment (melindaloubser1) [inline rasa/core/policies/ted_policy.py:1226]: What exception could be thrown here? Better to specify expected exceptions

**Review by melindaloubser1 — COMMENTED** [review]
I'll leave it to @kedz to provide a more thorough review, just noted two things to address for now

**Comment by melindaloubser1** [comment]
@SamuelNoB Could you please sign the contributor license agreement in the meantime?
@WashingtonBispo Do the comments left here so far make sense?

File: rasa/core/policies/ted_policy.py
Comment (dakshvar22) [inline rasa/core/policies/ted_policy.py:1222]: Why is this needed if you have done the correct device placement in `run_training`?

File: rasa/core/policies/ted_policy.py
Comment (ka-bu) [reply rasa/core/policies/ted_policy.py:638]: There are two ways do de-duplicate the code here:
1. We could put the code into a function
```
if condition:
  # don't use context manager
   run_code_that_needs_to_run_in_both_cases() 
else:
    with special_context_manager:  # i.e. cpu device placement
       run_code_that_needs_to_run_in_both_cases()
``` 
2. A nice way that avoids having to create another function is to use a [nullcontext](https://docs.python.org/3/library/contextlib.html#contextlib.nullcontext). With such a `nullcontext`, we can write ...
```
with (special_context_manager if condition else nullcontext):
   run_code_run_with_or_without_context_manager() 
``` 
   -  ... to achieve the same as in 1.
   - (Btw, `nullcontext`s are available for Python >= 3.7 and [luckily we dropped support for Python <= 3.6](https://github.com/RasaHQ/rasa/blob/main/pyproject.toml#L66) :)

File: rasa/core/policies/ted_policy.py
Comment (ka-bu) [inline rasa/core/policies/ted_policy.py:1230]: I think this might disable GPU use for everything that follows TED training - within the same python session - which would break cases where we run inference right after training (and expect GPU to be used for that).

Comment (WashingtonBispo) [reply rasa/core/policies/ted_policy.py]: Yeah, this is an old version i have tested. Sorry, I forgot to exclude.

File: rasa/core/policies/ted_policy.py
Comment (dakshvar22) [inline rasa/core/policies/ted_policy.py:345]: Lets keep this `True` because CPU being faster than GPU seems more like a bug and this should be set to `False` only as a workaround of this bug :) 
```suggestion
            USE_GPU: True,
```

File: rasa/core/policies/ted_policy.py
Comment (dakshvar22) [inline rasa/core/policies/ted_policy.py:714]:
```suggestion
        if self.config[USE_GPU]:
```

File: rasa/core/policies/unexpected_intent_policy.py
Comment (dakshvar22) [inline rasa/core/policies/unexpected_intent_policy.py:289]:
```suggestion
            USE_GPU: True,
```

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Thanks for the quick fixes. I have made some more suggestions + a couple of things that are pending:

1. We need to make sure that inference also runs on the requested device. In order to do that, we can wrap the [model loading call](https://github.com/RasaHQ/rasa/blob/ec8de9bc19c20880bf681a686f776523b612cc27/rasa/core/policies/ted_policy.py#L1076) with the appropriate context manager as you have done in `train`. The value of `USE_GPU` would have been persisted during training and will be loaded correctly in the `config` variable (line 1067), so you can use `config` to fetch the value of `USE_GPU`.

2. Please add a description of `use_gpu` to the table of parameters of [TEDPolicy](https://github.com/RasaHQ/rasa/blob/9fbd2877276b630ab36a99c487ea9af4b590141f/docs/docs/policies.mdx#L222) and [UnexpecTEDIntentPolicy](https://github.com/RasaHQ/rasa/blob/9fbd2877276b630ab36a99c487ea9af4b590141f/docs/docs/policies.mdx#L603)

3. Please add a changelog describing the change made in this PR to the [changelog folder](https://github.com/RasaHQ/rasa/tree/main/changelog). Please refer to the README inside that folder for instructions on the content of the changelog file.

**Comment by WashingtonBispo** [comment]
> 
 I'm not sure if i know how to make this first request.

File: pyproject.toml
Comment (ka-bu) [inline pyproject.toml:152]: Please revert this change (or explain why this is necessary)

**Review by ka-bu — COMMENTED** [review]
The following still needs to be implemented to make sure that we also switch to CPU when loading and finetuning an existing model: 

> We need to make sure that inference also runs on the requested device. In order to do that, we can wrap the [model loading call](https://github.com/RasaHQ/rasa/blob/ec8de9bc19c20880bf681a686f776523b612cc27/rasa/core/policies/ted_policy.py#L1076) with the appropriate context manager as you have done in train. The value of USE_GPU would have been persisted during training and will be loaded correctly in the config variable (line 1067), so you can use config to fetch the value of USE_GPU.

Some more details on that:
1. If we want to use the CPU for finetuning, then we need to wrap the `model = cls._load_tf_model(...)` call that Daksh referenced above into a `with tf.device(...)` as well (because that call will call the RasaModel load function which contains the compile that builds the graph for the loaded model and which will put it on the GPU if not told otherwise)
2. Since that call happens in a class function, we don't have access to `self.config` as before and need to get the information whether we want to use the GPU or not from somewhere else. We can get the same information from the `config` object that is created a few lines above the `model = cls._load_tf_model(...)`  call. 

Let me know if this helps or if you have more questions

File: pyproject.toml
Comment (dakshvar22) [reply pyproject.toml:152]: @ka-bu I think you reviewed the changes from last commit. [Comparing it with `main`, `pytest-timeout` should be left at `^1.4.2`](https://github.com/RasaHQ/rasa/pull/10944#discussion_r819591297).

**Comment by dakshvar22** [comment]
Hi @WashingtonBispo , could you please give an estimated timeline by when you could address the remaining comments? We would like to get this merged asap for an upcoming release. Thanks a lot for your contributions so far 🙏

**Comment by WashingtonBispo** [comment]
Sorry for being late, I will work on this weekend.

File: poetry.lock
Comment (ka-bu) [inline poetry.lock:2162]: Please revert all the changes in the lock file to the version that is currently in main

File: changelog/10897.improvement.md
Comment (ka-bu) [inline changelog/10897.improvement.md:1]:
```suggestion
Added flag to `TEDPolicy` that can be used to enable training on CPU. 
```

File: docs/docs/policies.mdx
Comment (ka-bu) [inline docs/docs/policies.mdx:216]:
```suggestion
  This parameter defines whether a GPU (if available) will be used training. By default, `TEDPolicy` will be trained on GPU if a GPU is available (i.e. `use_gpu` is `True`). To enforce that `TEDPolicy` uses only the CPU for training, set `use_gpu` to `False`.
```

File: rasa/core/policies/ted_policy.py
Comment (ka-bu) [inline rasa/core/policies/ted_policy.py:1085]: As explained in the previous comment, please avoid code duplication by using a nullcontext:
```suggestion
        with(tf.device('/cpu:0') if config['use_gpu'] else contextlib.nullcontext()):
            model = cls._load_tf_model(
                    model_utilities,
                    model_data_example,
                    predict_data_example,
                    featurizer,
                    execution_context.is_finetuning,
            )
```

File: rasa/core/policies/ted_policy.py
Comment (ka-bu) [inline rasa/core/policies/ted_policy.py:718]: As explained in the previous comments, please use a nullcontext to avoid code duplication:
```suggestion
        with( tf.device('/cpu:0') if self.config[USE_GPU] else self.config[USE_GPU]nullcontext()) ):
            self.run_training(model_data, label_ids)
```

File: rasa/core/policies/unexpected_intent_policy.py
Comment (ka-bu) [inline rasa/core/policies/unexpected_intent_policy.py:289]: Please extend the changelog and the readmes also for `UnexpecTEDIntentPolicy`.

**Review by ka-bu — COMMENTED** [review]
Looks good now :) Just some final finetuning needed

**Comment by dakshvar22** [comment]
@WashingtonBispo Can you please merge `main` into your branch and remove the changes you have made to `poetry.lock`?

**Comment by emysdias** [comment]
> @WashingtonBispo Can you please merge `main` into your branch and remove the changes you have made to `poetry.lock`?

Now it is fixed

File: rasa/core/policies/ted_policy.py
Comment (dakshvar22) [inline rasa/core/policies/ted_policy.py:1080]:
```suggestion
        model = None
```

File: changelog/10897.improvement.md
Comment (dakshvar22) [inline changelog/10897.improvement.md:1]:
```suggestion
Added flag `use_gpu` to `TEDPolicy` and `UnexpecTEDIntentPolicy` that can be used to enable training on CPU even when a GPU is available. 
```

**Comment by dakshvar22** [comment]
Hi @WashingtonBispo @emysdias Added a couple of final suggestions + the code quality check is failing above. Can you please address these?

**Comment by emysdias** [comment]
> Hi @WashingtonBispo @emysdias Added a couple of final suggestions + the code quality check is failing above. Can you please address these?

done

File: rasa/core/policies/unexpected_intent_policy.py
Comment (dakshvar22) [inline rasa/core/policies/unexpected_intent_policy.py:8]:
```suggestion
```

**Comment by dakshvar22** [comment]
Changes needed still

**Comment by emysdias** [comment]
> Changes needed still

ops, I'm going to change

**Comment by dakshvar22** [comment]
Running final model regression tests on this [PR](https://github.com/RasaHQ/rasa/pull/11023)

**Comment by dakshvar22** [comment]
hi @emysdias and @WashingtonBispo , all tests seem to be passing now. Can you please merge the base branch once more and then I can enable auto-merge 🎉

**Comment by emysdias** [comment]
> hi @emysdias and @WashingtonBispo , all tests seem to be passing now. Can you please merge the base branch once more and then I can enable auto-merge tada

done

**Review by dakshvar22 — APPROVED** [review]
Looks great! Thanks a lot for your contribution ✨

**Comment by emysdias** [comment]
> Looks great! Thanks a lot for your contribution sparkles

Thank you a lot

**Comment by WashingtonBispo** [comment]
> Looks great! Thanks a lot for your contribution ✨

Thank you sir.

**Pull request merged**
